### PR TITLE
Refactor/fix registering controllers

### PR DESCRIPTION
**Pull request by zpqrtbnk**
See #3593.

### CODE REVIEW

**Comment by lars-erik** [comment]
Looks better at a glance. Couple of quick Q's:

- Are TypeLoader something from LI, something in Umbraco, or some third-party thing?
- Could the collector pattern help? (sucky references, but here's the oldest one. Add ifs at leasure: http://wiki.c2.com/?CollectingParameterl)

File: src/Umbraco.Web/LightInjectExtensions.cs [hunk 1]
```diff
@@ -1,49 +1,82 @@
...
+            // We do NOT scan:
+            // - any IController or IHttpController (anything not PluginController nor UmbracoApiController), outside of Umbraco.Web
+            // which means that users HAVE to explicitly register their own non-Umbraco controllers
+            //
+            // This is because we try to achieve a balance between "simple" and "fast. Scanning for PluginController or
+            // UmbracoApiController is fast-ish because they both are IDiscoverable. Scanning for IController or IHttpController
+            // is a full, non-cached scan = expensive, we do it only for 1 assembly.
+            //
+            // TODO
+            // find a way to scan for IController *and* IHttpController in one single pass
```
Comment (JimBobSquarePants) [inline src/Umbraco.Web/LightInjectExtensions.cs]: Register them manually please.

File: src/Umbraco.Core/Composing/TypeFinder.cs [hunk 2]
```diff
@@ -210,53 +210,55 @@ internal static HashSet<Assembly> GetAllAssemblies()
...
-                    "itextsharp,",
-                    "UrlRewritingNet.",
-                    "HtmlAgilityPack,",
-                    "MiniProfiler,",
-                    "Moq,",
-                    "nunit.framework,",
-                    "TidyNet,",
-                    "WebDriver,"
-                };
+        internal static readonly string[] KnownAssemblyExclusionFilter = {
```
Comment (JimBobSquarePants) [inline src/Umbraco.Core/Composing/TypeFinder.cs]: This is an arms race. You'd be much better off registering your own assemblies and creating a simple means for consumers to register their own assemblies for discovery.

Comment (Shazwazza) [reply src/Umbraco.Core/Composing/TypeFinder.cs]: This list is old and mostly arbitrary - i mentioned on the other thread, we don't scan all assemblies, we only scan assemblies that have references to our own assemblies for the types being scanned, similar to how .net core scans assemblies for controllers.

**Comment by warrenbuckley** [comment]
All good - merging this in :)

**Pull request merged**